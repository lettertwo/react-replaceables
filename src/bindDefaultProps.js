import {PropTypes} from 'react';
import compose from 'recompose/compose';
import getContext from 'recompose/getContext';
import defaultProps from 'recompose/defaultProps';
import componentFromProp from 'recompose/componentFromProp';

const replacedContext = {replacedComponent: PropTypes.func.isRequired};


export default function bindDefaultProps(props) {
  const enhance = compose(defaultProps(props), getContext(replacedContext));
  return enhance(componentFromProp('replacedComponent'));
}
