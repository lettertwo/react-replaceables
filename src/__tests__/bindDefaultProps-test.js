/* eslint-env jest */
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import bindDefaultProps from '../bindDefaultProps';
import renderer from 'react-test-renderer';

describe('bindDefaultProps', () => {
  const ReplacedComponent = ({test}) => <span>{test}</span>; // eslint-disable-line react/prop-types

  class Context extends Component {
    getChildContext() {
      return {replacedComponent: ReplacedComponent};
    }
    render() {
      return this.props.children;
    }
  }

  Context.propTypes = {children: PropTypes.element};
  Context.childContextTypes = {replacedComponent: PropTypes.func};

  it('applies props to the replaced component', () => {
    const Test = bindDefaultProps({test: 'value'});
    const tree = renderer
      .create(
        <Context>
          <Test />
        </Context>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("doesn't clobber props passed in by the parent", () => {
    const Test = bindDefaultProps({test: 'value'});
    const tree = renderer
      .create(
        <Context>
          <Test test="value2" />
        </Context>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
