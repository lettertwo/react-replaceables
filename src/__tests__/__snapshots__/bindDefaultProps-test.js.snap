// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bindDefaultProps applies props to the replaced component 1`] = `
<span>
  value
</span>
`;

exports[`bindDefaultProps doesn't clobber props passed in by the parent 1`] = `
<span>
  value2
</span>
`;
